---
title: 'Assignment 7: PGS'
output:
  github_document:
    toc: true
    toc_depth: 4
---

# Assignment Overview

In this assignment we will learn about population stratification, imputation of genotypes, and using polygenic scores.  Polygenic scores (PGSs) can be useful for predicting disease susceptibility. In order to calculate PGSs, we need two things: GWAS summary statistics (including effect sizes), and genotypes. Most of the time, only a subset of a person's genotypes are actually measured (e.g. via SNP array), and so we must impute the rest using a matched population of fully genotyped individuals. This is the goal of Assignment 7.

Throughout the assignment we will be using a Mini Cohort that has genetic data and some phenotypic variables, together with the 1000 Genomes Project samples. Both datasets are in bfile plink format, which encompasses 3 files: *.bim, .bed and .fam* all the files can be located under the following path: */projects/bmeg/A7*


# Getting Ready 

In this assignment, you will use the plink tool extensively. A plink tutorial can be found here: https://zzz.bwh.harvard.edu/plink/tutorial.shtml

```{bash, eval=FALSE}
## Install plink1.9 onto your A1 conda environment:
conda install -c bioconda plink
```


# Genotyping Quality Control


## General QC

Before we can start working on the genetic data, we need to ensure that the quality is adequate. Thus, we are gonna check the following measuring for our MiniCohort:

   1. **SNP call rate:** The call rate represents the percentage of participants with non-missing data for that SNP. Removing variants with a call rate lower than 95% avoids potential wrong calls to be included in further analysis
   
   2. **Minor Allele Frequency:** The minor allele frequency (MAF) echoes the less common allele frequency across the population. The MAF estimates tend to be more accurate for higher MAFs and the population sample size the MAF was based on. If there are too few samples representing the rare-allele, is hard to distinguish between a true rare-allele and sequencing errors.
   
   3. **Sample call rate:** Similar to SNP call rate, it allows to filter out all samples exceeding 98% missing genetic variants out of all the  calls. 
   

```{bash, eval=FALSE}
## Using only one run of plink 1.9 (with different flags)
## 1. Filter out -SNPs- with more than 5% missingness
## 2. Filter out -variants- with less than 1% MAF
## 3. Filter out -samples- with more than 2% missingness
## 4. Create an output file in bfile format (which contains the bed, fam and bim files) for the MiniCohort QCed data
#?# Type the command you used below: - 3pt
plink --bfile /projects/bmeg/A7/Mini_cohort --geno 0.05 --maf 0.01 --mind 0.02 --make-bed --out filtered_mini

```


## Global Ancestry Investigation

In order to enhance imputation accuracy when dealing with ethnically diverse cohorts is important to understand the genetic ancestries of the cohort's participants. Knowing the ancestral populations will ensure that the most closely related population is used as a reference for the imputation. For instance, one would not want to impute haplotypes of an individual of Yoruban ancestry with a population of East Asians because many of the haplotypes will differ between the two ancestries, leading to imputing the wrong variants for the Yoruban person. Hence, we will analyze the global ancestry of our cohort using Principal Component Analysis (PCA). PCA is an unsupervised, unbiased way to reduce the complexity of multidimensional.

## a. PCA-specific QC

We first need to ensure that only the most informative genetic variants are used in the analysis. To do this, we will: 

   1. **Filter out high linkage disequilibrium (LD) regions:** Because high LD regions will add redundancy to the PCA (leading to these regions dominating top PCs), they need to be removed. 
   
   2. **LD pruning:** Similarly, LD causes redundancy even outside the particularly problematic high-LD regions. Thus, we will use LD-pruning to identify variants that are in LD, and select one per block.
   
```{bash, eval=FALSE}
## Using only one run of plink 1.9 (with different flags)
## 1. Filter out the high-LD regions contained in the --high_LD_regions_hg19.txt-- file, located in /projects/bmeg/A7/
## 2. Use the --indep-pairwise to do LD prunning with the following parameters:
## - Window size: 200, 
## - Variant Count: 100 
## - r^2: 0.2 
#?# Type the command you use to create the Mini Cohort PCA-QCed bfile below: - 1pt
plink --bfile filtered_mini --exclude /projects/bmeg/A7/high_LD_regions_hg19.txt --indep-pairwise 200 100 0.2 --out filtered_mini_ld 
 
## Use the output -.prune.in- file to extract only the informative variants and create a new bfile format (bed, fam and bim files) from:
## 1. The General MiniCohort QC bfile created before
## 2. The 1KGP_reference bfile located in /projects/bmeg/A7/
#?# Type the commands you used below: - 3pt
plink --bfile filtered_mini --exclude filtered_mini_ld.prune.in --make-bed --out mini_extract
plink --bfile /projects/bmeg/A7/1kgp_reference --exclude filtered_mini_ld.prune.in --make-bed --out 1kgp_extract

```

## b. PCA computation

To assess the ethnic diversity in our cohort, we will use One-thousand Genome Project (1KGP) data as a reference for our PCA analysis. These dataset has genetic information of major continental populations: Admixed American (AMR), European (EU), Asian (AS) and African (A). 

```{bash, eval=FALSE}
## Merge your pruned bfiles of the Mini_cohort and the 1KGP created on the previous step 
## Remember to create a new bfile (.fam, .bed and .bim files) that contains the merged data.
## IMPORTANT TIME CONSTRAINT: This step can take ~15 minutes, so make sure to check the server status before you run it!
#?# Type the command you used below: - 1pt
plink --bfile mini_extract --merge 1kgp_extract --make-bed --out merge_files

plink --bfile mini_extract --bmerge 1kgp_extract.bed 1kgp_extract.bim 1kgp_extract.fam --make-bed --recode --out merge_files

#?# Perform a PCA analysis in plink on the merged set - 1 pt
plink --bfile merge_files --pca

```

## c. Visualization

```{r}
library(ggplot2)
## Copy the PCA .eigenvec file to your computer, together with the samples_info.txt located in /projects/bmeg/A7/
## Load the .eigenvec file onto R, change the column names to: FID, IID, PC1, PC2, PC3, ..., PC20
#?# Type the command you used below: - 1pt
eigen_load <- read.table("plink.eigenvec", header = FALSE)

## Load the samples_info.txt file onto R, change the column names to: FID, IID, SuperPopulation, Population
#?# Type the commands you used below: - 1pt
samps_info <- read.table("samples_info.txt", header = TRUE)
colnames(samps_info) <- c("FID", "IID", "SuperPopulation", "Population")

## Merge the .eigenvec and sample_info data.frames together using the IID column
## Tip: Look into the -merge- function!
#?# Type the command you used below: - 1pt
merged_df <- merge(eigen_load, samps_info, by.x = "V2", by.y = "FID")

## Using ggplot create a scatterplot, using: 
## x-axis: PC1
## y-axis: PC2
## color: SuperPopulation - to use the Population information to color the samples and be able to appreciate population structure!
#?# Type the command you used below: 1pt
ggplot(data = merged_df, mapping = aes(x = V3, y = V4, colour = SuperPopulation)) + geom_point()


#?# Where do the cohort samples fall? Are they all clustered together? - 1 pt

#Based on the above plot, three main clusters can immediately be identified.
#The European (red), East Asian (green), and Eurpean (teal) cluster quite distinctly 
#from each other. The American (brown), Mini cohort (blue) and SAS (pink) populations 
#overlap with each other, but the SAS data appears to cluster more tightly
#than the america population.


#?# Which Population would you use as a reference for imputation?, Why? - 1 pt
# I would assume that I would want to use a population that encompasses a lot
# of variation (ie. doesn't cluster as tightly together on the PC plot) because 
# you would want to be able to best represent variation observed across multiple 
# populations. From the figure, we can see that AMR does not cluster as close 
# together, which indicates to me that it should be able to capture more of the 
# variation in the population, making it the best population to choose for data 
# imputation.


#?# Do you think looking at the top two PCs is sufficient to tell what population is best? Why/why not? - 2 pt
# I would say yes, because the majority of the variance between the population 
# can be extracted from the top two ranked PCs. This is backed up by looking at 
# the eigen values, which can be used to quantify the "amount" of variance that 
# is a result of each PC. From the eigen values calculated from plink, we can 
# see that PC1 and 2 have eigen values of 212.983 and 104.46 respectively. Compare 
# this to the other PC eigen values, which are much smaller (ranging from 29.8641
# to 1.83), indicating that the explain a much smaller proportion of the variance
# observed in the data. Therefore, I conclude that looking at the top two PCs
# are enough to examine what population is best for analysis


```

# Imputation

Imputation of genetic data is a very computationally intensive analysis, that can take a long time. So we have performed it for you. Using the chromosome 17 imputation information located in */projects/bmeg/A7/* under the *Mini_cohort_chr17_imputation_results.info.gz* we will calculate some post-imputation metrics. 

```{r, eval = FALSE}
## Load the Mini_cohort_chr17_imputation_results.info.gz file to your Rstudio environment 
#chr_info <- read.table(file = "/Users/paola.arguello/Documents/BMEDG400/Assignment_Keys/2021Jan/Assignment_7/Mini_cohort_chr17_imputation_results.info.gz",
 #                            header = TRUE)
## Use the information in the file to answer the following questions. Accompany each of the answers with the code you used to get to them and a brief explanation of your thought process behind.
#?# What is the percentage of imputed SNPs? 0.5 pt
## The metric of imputation quality is Rsq, this is the estimated value of the squared correlation between imputed and true genotypes. Since true genotypes are not available, this calculation is based on the idea that poorly imputed genotype counts will shrink towards their expectations based on allele frequencies observed in the population (https://genome.sph.umich.edu/wiki/Minimac3_Info_File#Rsq).  An Rsq < 0.3 is often used to flag poorly imputed SNPs. 
#?# What is the percentage of poorly imputed SNPs?
#?# Create a histogram to visualize the distribution of the MAF - 1 pt
#?# Which MAF is most frequent? What does that mean? - 1 pt
#?# What is the maximum MAF? Why is that? - 1 pt
```

# Polygenic Scores (PGS) 

A GWAS for affinity for tapas (the Spanish appetizer) was performed and 199 SNPs were found significantly associated. The significant SNPs and their assigned effect sizes are described in the *Tapas_enjoyability_GWAS_sumStats.txt* file. Thanks to the imputation performed in our MiniCohort, we were able to obtain the dosages (double risk alleles=2, one risk allele=1, no risk alleles=0) for each one of the SNPs associated to the Tapas 'enjoyability', described in the *MiniCohort_Tapas_SNPdosages.txt*. 

PGS are calculated by multiplying the effect sizes of each SNP by the dosage of an individual for those SNP and then adding together all the effectSize x dosage. The formula is outlined below, where:

  - i: individual of which you are calculating the PGS
  
  - j: SNP that has been found to be associated to the trait (Tapas enjoyability in this case)

  - Beta: Effect size

  - dosage: number of risk alleles the *individual i* has of the *risk allele j*? (2,1 or 0)

![](PGS_formula.png)

```{r}
## Load to your RStudio:
## 1.  -Tapas_enjoyability_GWAS_sumStats.txt-
## 2.  -MiniCohort_Tapas_SNPdosages.txt- 
## Both are located in the A7 directory on github.
## Using the base PRS formula outlined below, calculate the Tapas enjoyability PGS for the individuals in the Mini Cohort 
#?# Include your rationale and the documented code you used - 5pt
#?# Use ggplot to plot the distribution of the Tapas PGS: - 2 pt
## Include the code and the graph in your analysis! 
## Tip: http://www.cookbook-r.com/Graphs/Plotting_distributions_(ggplot2)/
 
#?# What is the distribution of the tapas PGS? - 1pt
```

## PGS accuracy

```{r}
## The Tapas enjoyability was measured in a range of 0-1, with 0 being hating tapas and 1 being completely in love with tapas.
## This tapas likability is captured in the "Tapas_enjoyability" column of the -MiniCohort_Tapas_SNPdosages.txt- file. 
#?# Make a scatterplot with a linear regression line, where x is the Tapas-PGS and y is their actual Tapas enjoyability - 2 pt
## Tip: http://www.sthda.com/english/wiki/ggplot2-scatter-plots-quick-start-guide-r-software-and-data-visualization
#?# What is the correlation coefficient between the PGS and Tapas enjoyability? Is Spearman or Pearson correlation more appropriate here? Why? - 3 pt
#?# How predictive is the PGS for tapas preference? Include in your answer why do you think it is/isn't accurate and what could affect its predicitvity - 2pt 
```


# Authors and contributions

Following completion of your assignment, please fill out this section with the authors and their contributions to the assignment. If you worked alone, only the author (e.g. your name and student ID) should be included.

Authors: Neera Patadia (79557773)

